File: scripts/verify_dic_manifest.py
#!/usr/bin/env python3
"""dic/MANIFEST.jsonが記錄するSHA-256を檢證する。"""

from __future__ import annotations

import hashlib
import json
import sys
from pathlib import Path


DATA_DIR = Path(__file__).resolve().parent.parent / "dic"
MANIFEST_PATH = DATA_DIR / "MANIFEST.json"


def main() -> int:
    manifest = json.loads(MANIFEST_PATH.read_text(encoding="utf-8"))
    invalid = False
    for name, expected in manifest["files"].items():
        path = DATA_DIR / name
        actual = hashlib.sha256(path.read_bytes()).hexdigest()
        if actual != expected:
            print(f"{name}: expected {expected}, got {actual}", file=sys.stderr)
            invalid = True
    return int(invalid)


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/verify_dic_manifest.clj
#!/usr/bin/env bb
(ns verify-dic-manifest
  "dic/MANIFEST.jsonに記錄されたSHA-256を檢證する。"
  (:require
    [babashka.fs :as fs]
    [cheshire.core :as json])
  (:import
    (java.math
      BigInteger)
    (java.security
      MessageDigest)))


(def data-dir
  (-> *file*
      fs/path
      fs/parent
      fs/parent
      (fs/path "dic")))


(def manifest-path (fs/path data-dir "MANIFEST.json"))


(defn sha256
  [path]
  (let [digest (MessageDigest/getInstance "SHA-256")]
    (format "%064x" (BigInteger. 1 (.digest digest (fs/read-all-bytes path))))))


(defn main
  []
  (let [manifest (json/parse-string (slurp (str manifest-path)))
        invalid? (atom false)]
    (doseq [[name expected] (get manifest "files")]
      (let [actual (sha256 (fs/path data-dir name))]
        (when (not= actual expected)
          (binding [*out* *err*]
            (println (format "%s: expected %s, got %s" name expected actual)))
          (reset! invalid? true))))
    (if @invalid? 1 0)))


(System/exit (main))
